watcher/watcher.go

```diff
@@ -600,75 +594,206 @@ func (w *Watcher) checkState(
-			scheduler.StateLastChangedAt = nowTimestamp
-			changedState = true
-		} else if nowTimestamp-scheduler.LastScaleOpAt > int64(autoScalingInfo.Up.Cooldown) {
-			shouldScaleUp = true
+
+	// Down
+	if scaling.Delta == 0 && scaling.InSync && roomCount.Total() > autoScalingInfo.Min {
+		if len(autoScalingInfo.Down.MetricsTrigger) > 0 {
+			scaling, err = w.checkMetricsTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
+		} else {
+			scaling, err = w.checkLegacyTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
 		}
 	}
 
-	threshold = autoScalingInfo.Down.Trigger.Threshold
-	usage = float32(autoScalingInfo.Down.Trigger.Usage) / 100
-	capacity = autoScalingInfo.Down.Trigger.Time / w.AutoScalingPeriod
+	if scaling.InSync && scheduler.State != models.StateInSync {
+		scheduler.State = models.StateInSync
+		scheduler.StateLastChangedAt = nowTimestamp
+		scaling.ChangedState = true
+	}
+
+	return scaling, err
+}
+
+func (w *Watcher) checkMetricsTrigger(
+	autoScalingInfo *models.AutoScaling,
+	scalingPolicy *models.ScalingPolicy,
+	scaleInfo *models.ScaleInfo,
+	roomCount *models.RoomsStatusCount,
+	scheduler *models.Scheduler,
+	nowTimestamp int64,
+) (*scaling, error) {
+	isDown := autoScalingInfo.Down == scalingPolicy
+	unbalancedState := models.StateSubdimensioned
+	scaleType := models.ScaleTypeUp
+	scaling := &scaling{
+		Delta:        0,
+		ChangedState: false,
+		InSync:       true,
+	}
 
-	isAboveThreshold, err = w.ScaleDownInfo.SendUsageAndReturnStatus(
+	if isDown {
+		scaleType = models.ScaleTypeDown
+		unbalancedState = models.StateOverdimensioned
+	}
+
+	for _, trigger := range scalingPolicy.MetricsTrigger {
+		threshold := trigger.Threshold
+		usage := float32(trigger.Usage) / 100
+		capacity := trigger.Time / w.AutoScalingPeriod
+
+		// Limit define a threshold that if surpassed should trigger scale up no matter what.
+		if !isDown {
+			isAboveLimit, err := w.checkIfUsageIsAboveLimit(trigger, roomCount, scheduler, scaling, unbalancedState, trigger.Limit, nowTimestamp)
+			if err != nil {
+				return scaling, err
+			}
+			if isAboveLimit {
+				return scaling, nil
+			}
+		}
+
+		isAboveThreshold, err := scaleInfo.ReturnStatus(
+			w.SchedulerName,
+			trigger.Metric,
+			scaleType,
+			capacity, roomCount.Total(),
+			threshold, usage,
+		)
+		if err != nil {
+			return scaling, err
+		}
+
+		if isAboveThreshold {
+			scaling.InSync = false
+			if scheduler.State != unbalancedState {
+				scaling.ChangedState = true
+				scheduler.State = unbalancedState
+				scheduler.StateLastChangedAt = nowTimestamp
+			}
+
+			// changed state or not in cooldown window
+			if scaling.ChangedState || nowTimestamp-scheduler.LastScaleOpAt > int64(scalingPolicy.Cooldown) {
+				scaling.Delta, err = dynamicDelta(trigger, roomCount)
+				return scaling, err
+			}
+		}
+	}
+
+	return scaling, nil
+}
+
+func (w *Watcher) checkLegacyTrigger(
```

Comment:
There is too much duplicated code with checkMetricsTrigger. Couldn't you merge them? For example, The for loop on checkMetricsTrigger could be extracted and receive the same parameters checkLegacyTrigger uses.

Reply:
will do